state/spaces.go

```diff
@@ -529,13 +530,96 @@ func (st *State) SaveProviderSpaces(providerSpaces []network.SpaceInfo) error {
 			}
 
 			spaceNames.Add(spaceName)
-			spaceId = space.Id()
+			spaceID = space.Id()
+
+			// To ensure that we can remove spaces, we backfill the new spaces
+			// onto the modelSpaceMap.
+			modelSpaceMap[space.ProviderId()] = space
 		}
 
-		err = st.SaveProviderSubnets(spaceInfo.Subnets, spaceId)
+		err = st.SaveProviderSubnets(spaceInfo.Subnets, spaceID)
 		if err != nil {
 			return errors.Trace(err)
 		}
+
+		updatedSpaces.Add(spaceInfo.ProviderId)
+	}
+
+	// Workout the difference between all the current spaces vs what was
+	// actually changed.
+	allStateSpaces := network.MakeSubnetSet()
+	for providerID := range modelSpaceMap {
+		allStateSpaces.Add(providerID)
+	}
+
+	defaultEndpointBinding, err := st.DefaultEndpointBindingSpace()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	model, err := st.Model()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	allEndpointBindings, err := model.AllEndpointBindings()
+	if err != nil {
+		return errors.Trace(err)
 	}
+
+	allEndpointBindingsSpaces := set.NewStrings()
+	for _, binding := range allEndpointBindings {
+		bindingSpaceNames, err := binding.Bindings.MapWithSpaceNames()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, spaceName := range bindingSpaceNames {
+			allEndpointBindingsSpaces.Add(spaceName)
+		}
+	}
+
+	remnantSpaces := allStateSpaces.Difference(updatedSpaces)
+	for _, providerID := range remnantSpaces.SortedValues() {
```

Comment:
I know we are in state here, but can we log at each of the continue points somehow?
Maybe pass something out to the caller if we don't want to log directly.